archive/untar_stream.ts

```diff
@@ -0,0 +1,352 @@
+   */
+  linkname: string;
+  /**
+   * The magic number of the entry.
+   */
+  magic: string;
+  /**
+   * The version number of the entry.
+   */
+  version: string;
+  /**
+   * The uname of the entry.
+   */
+  uname: string;
+  /**
+   * The gname of the entry.
+   */
+  gname: string;
+  /**
+   * The devmajor of the entry.
+   */
+  devmajor: string;
+  /**
+   * The devminor of the entry.
+   */
+  devminor: string;
+  /**
+   * The former half of the name of the entry.
+   */
+  prefix: string;
+}
+
+/**
+ * The header of an entry in the archive.
+ */
+export interface TarStreamHeader {
+  /**
+   * The type 'header' indicating the start of a new entry.
+   */
+  type: "header";
+  /**
+   * The pathname of the entry.
+   */
+  pathname: string;
+  /**
+   * The header of the entry.
+   */
+  header: OldStyleFormat | PosixUstarFormat;
+}
+
+/**
+ * The data belonging to the last entry returned.
+ */
+export interface TarStreamData {
+  /**
+   * The type 'data' indicating a chunk of content from the last 'header'
+   * resolved.
+   */
+  type: "data";
+  /**
+   * A chunk of content of from the entry.
+   */
+  data: Uint8Array;
+}
+
+/**
+ * The type extracted from the archive.
+ */
+export type TarStreamChunk = TarStreamHeader | TarStreamData;
+
+/**
+ * ### Overview
+ * A TransformStream to expand a tar archive.  Tar archives allow for storing
+ * multiple files in a single file (called an archive, or sometimes a tarball).
+ *   These archives typically have a single '.tar' extension.  This
+ * implementation follows the [FreeBSD 15.0](https://man.freebsd.org/cgi/man.cgi?query=tar&sektion=5&apropos=0&manpath=FreeBSD+15.0-CURRENT) spec.
+ *
+ * ### Supported File Formats
+ * Only the ustar file format is supported.  This is the most common format.
+ *   Additionally the numeric extension for file size.
+ *
+ * ### Usage
+ * When expanding the archive, as demonstrated in the example, one must decide
+ * to either consume the Readable Stream, if present, or cancel it. The next
+ * entry won't be resolved until the previous ReadableStream is either consumed
+ * or cancelled.
+ *
+ * ### Understanding Compressed
+ * A tar archive may be compressed, often identified by an additional file
+ * extension, such as '.tar.gz' for gzip. This TransformStream does not support
+ * decompression which must be done before expanding the archive.
+ *
+ * @example Usage
+ * ```ts no-eval
+ * import { UnTarStream } from "@std/archive/untar-stream";
+ *
+ * let fileWriter: WritableStreamDefaultWriter | undefined;
+ * for await (
+ *   const entry of (await Deno.open('./out.tar.gz'))
+ *     .readable
+ *     .pipeThrough(new DecompressionStream('gzip'))
+ *     .pipeThrough(new UnTarStream())
+ * ) {
+ *   if (entry.type === "header") {
+ *     fileWriter?.close();
+ *     fileWriter = (await Deno.create(entry.pathname)).writable.getWriter();
+ *   } else await fileWriter!.write(entry.data);
+ * }
+ * ```
+ */
+export class UnTarStream
```

Comment:
This should be renamed `UntarStream` ([untar](https://en.wiktionary.org/wiki/untar#English) is one word, not two words).